docs/proposals/maintenance.md

```diff
@@ -0,0 +1,179 @@
+
+## Overview
+
+Nodes need periodic maintenance, most of which require partial or total
+unavailability to pods.  The cluster admin would like for this to happen (1)
+automatically, and (2) respecting any service quality constraints.
+
+Examples of maintenance:
+
+- unmount/remount filesystems to change mkfs params
+- remove from service but leave up for debugging
+- reboot for kernel or image upgrade
+- remove from service for hardware repair
+- remove from service permanently
+
+## Service Quality
+
+This may be thought of as the cluster admin offering a guarantee or SLO, or as the
+user specifying their requirements.  Around evictions, there are typically three
+parameters:
+
+- time between evictions (of pods in the same replica set)
+- minimum shard strength (minimum fraction of pods healthy)
+- eviction notice
+
+## Implementations
+
+In order of increasing complexity (and benefit).
+
+### Drain one node
+
+This is implemented in https://github.com/kubernetes/kubernetes/pull/16698
+
+### One-shot drain of multiple nodes
+
+In this case, we try to meet some specific service goals specified by the admin
+on the command line.  The specification of these parameters only when we run the
+tool lets us provide the possibility of safety while avoiding questions of where
+to configure and store such specification (treated further down).
+
+```
+kubectl drain \
+  --min-shard-strength=0.75 \
+  --min-seconds-between-evictions=900 \
+  --grace=900 $nodes
+```
+
+This would still be pretty simple.  We offer no guarantees about safety in the
+face of crashes or restarts, but we do meet the needs of admins who want
+something more sophisticated than `sleep`.
+
+### Implementation of the above as a controller
+
+The parameters of a specific lame, drain, or unlame request could be pushed to
+the API server as a drain object.  The drain controller would have to manage
+multiple drains in flight at once.  If the drains have differing parameters,
+then more aggressive drains would generally starve safer ones.  While this could
+be a useful feature in emergencies, it will be a lot easier to reason about
+multiple drains in progress if they all share the same parameters, which would
+be configured via kube-env and would represent the cluster admin's SLO.
+
+The drain object would thus look something like this
+
+```
+type DrainSpec struct {
+  LameNodes    PodList
+  RemoveNodes  PodList
+  AddNodes     PodList
+  GracePeriod  int
+}
+```
+
+A drain object would work much like a deployment.  A drain controller would
+watch for drain objects and, within constraints, work toward the desired end
+state.  Nodes in `LameNodes` would merely be lamed, nodes is `RemoveNodes` would
+be drained according to the given parameters, and nodes in `AddNodes` would be
+undrained.  The object would be rejected if the sets overlap at all.  The action
+of the drain controller can be paused or stopped by deleting the drain object,
+and restarted by re-posting it.
```

Comment:
restarting seems to require some way of knowing which machines are already in drained state (we already know which are in lamed state, namely the unschedulable flag, though I'm a little concerned we're overloading it), which gets back to the earlier issue of does a machine signal it is in drained state via status, or phase, or spec...